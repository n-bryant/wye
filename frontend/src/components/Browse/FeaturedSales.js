import React from "react";
import PropTypes from "prop-types";

import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../../lib/classNamePrefix";
import { withStyles } from "@material-ui/core/styles";
import { withWidth } from "@material-ui/core/";

import Grid from "@material-ui/core/Grid";
import Box from "@material-ui/core/Box";

import QuickLink from "../QuickLink";
import PaginationWidget from "../PaginationWidget";
import FeaturedWidget from "./FeaturedWidget";
import BrowseCard from "./BrowseCard";
import styles from "./FeaturedSales.styles";

export const TITLE = "Sales";
export const SUBTITLE = "Big sales for some of the hottest titles";

/**
 * renders a FeaturedWidget with cards representing some of the best sales
 */
export const FeaturedSales = props => {
  const classnames = FeaturedSales.classnames(props);
  const { items, width } = props;

  // pagination values
  const perPage = 3;
  const [currentPage, setCurrentPage] = React.useState(1);
  const pageStart =
    currentPage === 1 ? 0 : items.length - (currentPage - 1) * perPage;
  const pageEnd = currentPage * perPage;

  // divide items into featured/sub-featured
  const featuredSales = items.slice(0, 2);
  const subFeaturedSales = items.slice(2, 6);

  return (
    <FeaturedWidget
      className={classnames.root()}
      title={TITLE}
      subTitle={SUBTITLE}
    >
      {!["xs", "sm"].some(val => val === width) ? (
        <Grid
          className={classnames.element("featuredSaleContainer")}
          container
          spacing={2}
        >
          <Grid
            className={classnames.element("subFeaturedSaleContainer")}
            container
            item
            spacing={2}
            xs={6}
          >
            {subFeaturedSales.map(saleItem => (
              <Grid
                className={classnames.element("subFeaturedSale")}
                key={saleItem.appid}
                item
                xs={6}
              >
                <BrowseCard data={saleItem} variant="header" withPrice={true} />
              </Grid>
            ))}
          </Grid>
          <Grid container item spacing={2} xs={6}>
            {featuredSales.map(saleItem => (
              <Grid
                className={classnames.element("featuredSale")}
                key={saleItem.appid}
                item
                xs={6}
              >
                <BrowseCard data={saleItem} variant="lib" withPrice={true} />
              </Grid>
            ))}
          </Grid>
        </Grid>
      ) : (
        <React.Fragment>
          <Grid
            className={classnames.element("featuredSaleContainer")}
            container
            spacing={2}
          >
            {items.slice(pageStart, pageEnd).map(saleItem => (
              <Grid
                className={classnames.element("featuredSale")}
                key={saleItem.appid}
                item
                xs={4}
              >
                <BrowseCard data={saleItem} variant="lib" withPrice={true} />
              </Grid>
            ))}
          </Grid>
          <PaginationWidget
            className={classnames.element("paginationWidget")}
            currentPage={currentPage}
            setCurrentPage={setCurrentPage}
            totalPages={items.length / perPage}
          />
        </React.Fragment>
      )}
      <Box className={classnames.element("quickLinksContainer")} my={4}>
        <QuickLink
          className={classnames.element("quickLink")}
          label={"under $5"}
          linkHref="/browse/under5"
        />
        <QuickLink
          className={classnames.element("quickLink")}
          label={"under $10"}
          linkHref="/browse/under10"
        />
      </Box>
    </FeaturedWidget>
  );
};
FeaturedSales.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}FeaturedSales`
);
FeaturedSales.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    featuredSaleContainer: PropTypes.string,
    featuredSale: PropTypes.string,
    subFeaturedSaleContainer: PropTypes.string,
    subFeaturedSale: PropTypes.string,
    paginationWidget: PropTypes.string,
    quickLinksContainer: PropTypes.string,
    quickLink: PropTypes.string
  }),
  // data for the cards to be rendered in the FeaturedWidget
  items: PropTypes.array,
  // material-ui width
  width: PropTypes.string
};
FeaturedSales.defaultProps = {
  classes: {}
};

// apply styles
export const StyledFeaturedSales = withWidth()(
  withStyles(styles)(FeaturedSales)
);

export default StyledFeaturedSales;
