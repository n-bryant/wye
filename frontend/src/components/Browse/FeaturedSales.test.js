import { FeaturedSales } from "./FeaturedSales";

describe("FeaturedSales", () => {
  const props = {
    classes: {
      root: "root",
      featuredSaleContainer: "featuredSaleContainer",
      featuredSale: "featuredSale",
      subFeaturedSaleContainer: "subFeaturedSaleContainer",
      subFeaturedSale: "subFeaturedSale",
      paginationWidget: "paginationWidget",
      quickLinksContainer: "quickLinksContainer",
      quickLink: "quickLink",
      spacer: "spacer"
    },
    items: [
      {
        appid: "1",
        name: "name",
        developers: ["a"],
        publishers: ["a"],
        ownersFormatted: "10,000 .. 20,000",
        userRating: 95,
        genres: ["a"],
        freeToPlay: false,
        onSale: true,
        discount: 25,
        initialPrice: 1000,
        finalPrice: 750,
        headerImage: "header",
        libraryCapsule: "lib"
      },
      {
        appid: "2",
        name: "name",
        developers: ["a"],
        publishers: ["a"],
        ownersFormatted: "10,000 .. 20,000",
        userRating: 95,
        genres: ["a"],
        freeToPlay: false,
        onSale: true,
        discount: 25,
        initialPrice: 1000,
        finalPrice: 750,
        headerImage: "header",
        libraryCapsule: "lib"
      },
      {
        appid: "3",
        name: "name",
        developers: ["a"],
        publishers: ["a"],
        ownersFormatted: "10,000 .. 20,000",
        userRating: 95,
        genres: ["a"],
        freeToPlay: false,
        onSale: true,
        discount: 25,
        initialPrice: 1000,
        finalPrice: 750,
        headerImage: "header",
        libraryCapsule: "lib"
      },
      {
        appid: "4",
        name: "name",
        developers: ["a"],
        publishers: ["a"],
        ownersFormatted: "10,000 .. 20,000",
        userRating: 95,
        genres: ["a"],
        freeToPlay: false,
        onSale: true,
        discount: 25,
        initialPrice: 1000,
        finalPrice: 750,
        headerImage: "header",
        libraryCapsule: "lib"
      },
      {
        appid: "5",
        name: "name",
        developers: ["a"],
        publishers: ["a"],
        ownersFormatted: "10,000 .. 20,000",
        userRating: 95,
        genres: ["a"],
        freeToPlay: false,
        onSale: true,
        discount: 25,
        initialPrice: 1000,
        finalPrice: 750,
        headerImage: "header",
        libraryCapsule: "lib"
      },
      {
        appid: "6",
        name: "name",
        developers: ["a"],
        publishers: ["a"],
        ownersFormatted: "10,000 .. 20,000",
        userRating: 95,
        genres: ["a"],
        freeToPlay: false,
        onSale: true,
        discount: 25,
        initialPrice: 1000,
        finalPrice: 750,
        headerImage: "header",
        libraryCapsule: "lib"
      }
    ],
    width: "xs"
  };

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<FeaturedSales {...props} />);
  });

  it("should render successfully", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should adjust the presentation of the cards if the width is md+", () => {
    wrapper.setProps({
      width: "md"
    });
    wrapper.update();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
