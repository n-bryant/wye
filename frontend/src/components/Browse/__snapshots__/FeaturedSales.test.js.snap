// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeaturedSales should adjust the presentation of the cards if the width is md+ 1`] = `
<WithStyles(FeaturedWidget)
  className="wye-FeaturedSales root"
  classes={Object {}}
  subTitle="Big sales for some of the hottest titles"
  title="Sales"
>
  <WithStyles(ForwardRef(Grid))
    className="wye-FeaturedSales__featuredSaleContainer featuredSaleContainer"
    container={true}
    spacing={2}
  >
    <WithStyles(ForwardRef(Grid))
      container={true}
      item={true}
      spacing={2}
      xs={6}
    >
      <WithStyles(ForwardRef(Grid))
        className="wye-FeaturedSales__featuredSale featuredSale"
        item={true}
        key="1"
        xs={6}
      >
        <HighlightTrailerQuery
          data={
            Object {
              "appid": "1",
              "developers": Array [
                "a",
              ],
              "discount": 25,
              "finalPrice": 750,
              "freeToPlay": false,
              "genres": Array [
                "a",
              ],
              "headerImage": "header",
              "initialPrice": 1000,
              "libraryCapsule": "lib",
              "name": "name",
              "onSale": true,
              "ownersFormatted": "10,000 .. 20,000",
              "publishers": Array [
                "a",
              ],
              "userRating": 95,
            }
          }
          variant="lib"
          withPrice={true}
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="wye-FeaturedSales__featuredSale featuredSale"
        item={true}
        key="2"
        xs={6}
      >
        <HighlightTrailerQuery
          data={
            Object {
              "appid": "2",
              "developers": Array [
                "a",
              ],
              "discount": 25,
              "finalPrice": 750,
              "freeToPlay": false,
              "genres": Array [
                "a",
              ],
              "headerImage": "header",
              "initialPrice": 1000,
              "libraryCapsule": "lib",
              "name": "name",
              "onSale": true,
              "ownersFormatted": "10,000 .. 20,000",
              "publishers": Array [
                "a",
              ],
              "userRating": 95,
            }
          }
          variant="lib"
          withPrice={true}
        />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="wye-FeaturedSales__subFeaturedSaleContainer subFeaturedSaleContainer"
      container={true}
      item={true}
      spacing={2}
      xs={6}
    >
      <WithStyles(ForwardRef(Grid))
        className="wye-FeaturedSales__subFeaturedSale subFeaturedSale"
        item={true}
        key="3"
        xs={6}
      >
        <HighlightTrailerQuery
          data={
            Object {
              "appid": "3",
              "developers": Array [
                "a",
              ],
              "discount": 25,
              "finalPrice": 750,
              "freeToPlay": false,
              "genres": Array [
                "a",
              ],
              "headerImage": "header",
              "initialPrice": 1000,
              "libraryCapsule": "lib",
              "name": "name",
              "onSale": true,
              "ownersFormatted": "10,000 .. 20,000",
              "publishers": Array [
                "a",
              ],
              "userRating": 95,
            }
          }
          variant="header"
          withPrice={true}
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="wye-FeaturedSales__subFeaturedSale subFeaturedSale"
        item={true}
        key="4"
        xs={6}
      >
        <HighlightTrailerQuery
          data={
            Object {
              "appid": "4",
              "developers": Array [
                "a",
              ],
              "discount": 25,
              "finalPrice": 750,
              "freeToPlay": false,
              "genres": Array [
                "a",
              ],
              "headerImage": "header",
              "initialPrice": 1000,
              "libraryCapsule": "lib",
              "name": "name",
              "onSale": true,
              "ownersFormatted": "10,000 .. 20,000",
              "publishers": Array [
                "a",
              ],
              "userRating": 95,
            }
          }
          variant="header"
          withPrice={true}
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="wye-FeaturedSales__subFeaturedSale subFeaturedSale"
        item={true}
        key="5"
        xs={6}
      >
        <HighlightTrailerQuery
          data={
            Object {
              "appid": "5",
              "developers": Array [
                "a",
              ],
              "discount": 25,
              "finalPrice": 750,
              "freeToPlay": false,
              "genres": Array [
                "a",
              ],
              "headerImage": "header",
              "initialPrice": 1000,
              "libraryCapsule": "lib",
              "name": "name",
              "onSale": true,
              "ownersFormatted": "10,000 .. 20,000",
              "publishers": Array [
                "a",
              ],
              "userRating": 95,
            }
          }
          variant="header"
          withPrice={true}
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="wye-FeaturedSales__subFeaturedSale subFeaturedSale"
        item={true}
        key="6"
        xs={6}
      >
        <HighlightTrailerQuery
          data={
            Object {
              "appid": "6",
              "developers": Array [
                "a",
              ],
              "discount": 25,
              "finalPrice": 750,
              "freeToPlay": false,
              "genres": Array [
                "a",
              ],
              "headerImage": "header",
              "initialPrice": 1000,
              "libraryCapsule": "lib",
              "name": "name",
              "onSale": true,
              "ownersFormatted": "10,000 .. 20,000",
              "publishers": Array [
                "a",
              ],
              "userRating": 95,
            }
          }
          variant="header"
          withPrice={true}
        />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(FeaturedWidget)>
`;

exports[`FeaturedSales should render successfully 1`] = `
<WithStyles(FeaturedWidget)
  className="wye-FeaturedSales root"
  classes={Object {}}
  subTitle="Big sales for some of the hottest titles"
  title="Sales"
>
  <WithStyles(ForwardRef(Grid))
    className="wye-FeaturedSales__featuredSaleContainer featuredSaleContainer"
    container={true}
    spacing={2}
  >
    <WithStyles(ForwardRef(Grid))
      className="wye-FeaturedSales__featuredSale featuredSale"
      item={true}
      key="1"
      xs={4}
    >
      <HighlightTrailerQuery
        data={
          Object {
            "appid": "1",
            "developers": Array [
              "a",
            ],
            "discount": 25,
            "finalPrice": 750,
            "freeToPlay": false,
            "genres": Array [
              "a",
            ],
            "headerImage": "header",
            "initialPrice": 1000,
            "libraryCapsule": "lib",
            "name": "name",
            "onSale": true,
            "ownersFormatted": "10,000 .. 20,000",
            "publishers": Array [
              "a",
            ],
            "userRating": 95,
          }
        }
        variant="lib"
        withPrice={true}
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="wye-FeaturedSales__featuredSale featuredSale"
      item={true}
      key="2"
      xs={4}
    >
      <HighlightTrailerQuery
        data={
          Object {
            "appid": "2",
            "developers": Array [
              "a",
            ],
            "discount": 25,
            "finalPrice": 750,
            "freeToPlay": false,
            "genres": Array [
              "a",
            ],
            "headerImage": "header",
            "initialPrice": 1000,
            "libraryCapsule": "lib",
            "name": "name",
            "onSale": true,
            "ownersFormatted": "10,000 .. 20,000",
            "publishers": Array [
              "a",
            ],
            "userRating": 95,
          }
        }
        variant="lib"
        withPrice={true}
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="wye-FeaturedSales__featuredSale featuredSale"
      item={true}
      key="3"
      xs={4}
    >
      <HighlightTrailerQuery
        data={
          Object {
            "appid": "3",
            "developers": Array [
              "a",
            ],
            "discount": 25,
            "finalPrice": 750,
            "freeToPlay": false,
            "genres": Array [
              "a",
            ],
            "headerImage": "header",
            "initialPrice": 1000,
            "libraryCapsule": "lib",
            "name": "name",
            "onSale": true,
            "ownersFormatted": "10,000 .. 20,000",
            "publishers": Array [
              "a",
            ],
            "userRating": 95,
          }
        }
        variant="lib"
        withPrice={true}
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(PaginationWidget)
    className="wye-FeaturedSales__paginationWidget paginationWidget"
    classes={Object {}}
    currentPage={1}
    setCurrentPage={[Function]}
    totalPages={2}
  />
</WithStyles(FeaturedWidget)>
`;
